### ISSUE TITLE
torch.dtype is int in JIT instead of torch.dtype

### ISSUE THREAD

**Issue by hal-314** [pytorch/pytorch#65607]
## 🐛 Bug

torch.dtype is int in JIT instead of torch.dtype. So, you loss to [properties: `is_floating_point` and `is_complex`](https://pytorch.org/docs/stable/tensor_attributes.html?highlight=dtype#torch.torch.dtype).

## To Reproduce

```python
import torch

@torch.jit.script
def is_floating_point(dtype: torch.dtype):
     return dtype.is_floating_point
``` 

Error (from similar methods
```
'int' object has no attribute or method 'is_floating_point'.:
  File "/home/victor/PycharmProjects/kornia/kornia/geometry/bbox_v2.py", line 615
        """Similar to `torch.nn.Module.to()` method"""
        # In torchscript, dtype is a int and not a class. https://github.com/pytorch/pytorch/issues/51941
        if dtype is not None and not dtype.is_floating_point:# _is_floating_point_dtype(dtype):
                                     ~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
            raise ValueError("Boxes must be in floating point")
        self._boxes = self._boxes.to(device=device, dtype=dtype)
```

## Expected behavior

JIT should keep torch.dtype methods so you avoid to implement a workaround to support JIT. If not, it should be clearly documented in docs. For example, in [TorchScript Unsupported Pytorch Constructs](https://pytorch.org/docs/stable/jit_unsupported.html#jit-unsupported)

## Environment
 Pytorch 1.9

**Comment by t-vi** [comment]
I'm not sure how to handle the BC breaking nature of changing the signature, but I'm working on a private branch that maps torch.dtype to the (existing (!), for two years) ScalarTypeType JIT type.

**Comment by t-vi** [comment]
My preliminary patch to ScriptTorch could be used as a POC what it takes to properly support dtypes (ONNX isn't updated yet, possibly also the fusers, but other than that, I'm relatively happy):

https://github.com/t-vi/pytorch/commit/d420f3527497a2dbc6211802d5e8d994918a8d3f

I should say that I have no ideas about how to deal with the backward compat breaking nature of the change other than bumping some serialization version and refuse to work with the old stuff (how would I do the versioning for my own branch)?

Probably some cleanups could be had from not needing to pretend around torch.dtype.

**Comment by t-vi** [comment]
@suo what should I do to move this forward? send a draft PR?

**Comment by hzhangxyz** [comment]
It there any update on this issue? I have the same problem here. I create a torch scripted function with parameter type as `torch.dtype`, but I can not pass correct dtype into this function.

**Comment by t-vi** [comment]
I don't think expect this will be fixed.

**Comment by hzhangxyz** [comment]
Oh :(. Is there any other way to get this problem work? It seems dtype in pytorch script is int, but in python side, It is dtype instead of int directly. It seems dtype is an enum inside pytorch, so is there any function to convert dtype to its internal int manually in python side to let torch script works?

**Comment by domef** [comment]
The only way that i found to know the mapping between dtype and corresponding int is this:

```python
import torch


class Cursed(torch.nn.Module):
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        print(torch.uint8)
        print(torch.int8)
        print(torch.int16)
        print(torch.int32)
        print(torch.int64)
        print(torch.float16)
        print(torch.float32)
        print(torch.float64)
        print(torch.complex64)
        print(torch.complex128)
        print(torch.bool)
        print(torch.bfloat16)
        return x


model = Cursed()
model_scripted = torch.jit.script(model)

x = torch.randn(4, 3, 224, 224)

model(x)
model_scripted(x)
```

The outcome is:

```
torch.uint8 ------> 0
torch.int8 -------> 1
torch.int16 ------> 2
torch.int32 ------> 3
torch.int64 ------> 4
torch.float16 ----> 5
torch.float32 ----> 6
torch.float64 ----> 7
torch.complex64 --> 9
torch.complex128 -> 10
torch.bool -------> 11
torch.bfloat16 ---> 15
```

I'm using  `torch==2.1.0+cu118` with `Ubuntu 22.04.5 LTS`.

**Comment by domef** [comment]
any update?

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
